lib/src/solver/version_solver.dart

```diff
@@ -424,17 +425,21 @@ class VersionSolver {
   /// The version list may not always be complete. If the package is the root
   /// package, or if it's a package that we didn't unlock while solving because
   /// we weren't trying to upgrade it, we will just know the current version.
-  Map<String, List<Version>> _getAvailableVersions(List<PackageId> packages) {
+  Future<Map<String, List<Version>>> _getAvailableVersions(
+      List<PackageId> packages) async {
     var availableVersions = <String, List<Version>>{};
     for (var package in packages) {
       var cached = _packageListers[package.toRef()]?.cachedVersions;
-      // If the version list was never requested, just use the one known
-      // version.
-      var versions = cached == null
-          ? [package.version]
-          : cached.map((id) => id.version).toList();
-
-      availableVersions[package.name] = versions;
+      // If the version list was never requested, use versions from cached
+      // version listings if the package is "hosted".
+      var ids = cached ??
+          (package.source is HostedSource
+              ? (await _systemCache
+                  .source(package.source)
+                  .getVersions(package.toRef(), maxAge: Duration(days: 3)))
+              : [package]);
```

Comment:
Does this really have to be `HostedSource` specific? Will git sources try to fetch something?

Reply:
Since this is only place we use `maxAge`, and we still have to use a `HostedSource` specific variant, isn't it kind of fair to ask if this is the right abstraction?

Reply:
Yeah - you are spot on. I added a TODO to find a better abstraction.